pkg/controller/clusterrelocate/clusterrelocate_controller.go

```diff
@@ -0,0 +1,797 @@
+		return reconcile.Result{}, nil
+	}
+
+	return r.finishRelocateCompletion(cd, desiredRelocate.Name, logger)
+}
+
+// setRelocateAnnotation sets the relocate annotation on the ClusterDeployment as well as on the child DNSZone, if there
+// is one.
+func (r *ReconcileClusterRelocate) setRelocateAnnotation(cd *hivev1.ClusterDeployment, relocateName string, status hivev1.RelocateStatus, logger log.FieldLogger) error {
+	dnsZone, err := r.dnsZone(cd, logger)
+	if err != nil {
+		return errors.Wrap(err, "failed to get DNSZone")
+	}
+
+	var objects []hivev1.MetaRuntimeObject
+	if dnsZone != nil {
+		// If setting to complete or clearing, do the DNSZone first. This ensures that if there is an error setting the
+		// annotation on the DNSZone, that it will be re-attempting on the next reconcile. If the ClusterDeployment is set
+		// first, then the next reconcile will not see that there is anything that needs to be done since the
+		// ClusterDeployment will already have the annotation in the right state.
+		if status == hivev1.RelocateComplete {
+			objects = []hivev1.MetaRuntimeObject{dnsZone, cd}
+		} else {
+			objects = []hivev1.MetaRuntimeObject{cd, dnsZone}
+		}
+	} else {
+		objects = []hivev1.MetaRuntimeObject{cd}
+	}
+
+	for _, obj := range objects {
+		if changed := controllerutils.SetRelocateAnnotation(obj, relocateName, status); !changed {
+			continue
+		}
+		if err := r.Update(context.Background(), obj); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "failed to set relocate annotation")
+			return errors.Wrap(err, "failed to set relocate annotation")
+		}
+	}
+	return nil
+}
+
+// clearRelocateAnnotation deletes the relocate annotation from the ClusterDeployment as well as on the child DNSZone,
+// if there is one.
+func (r *ReconcileClusterRelocate) clearRelocateAnnotation(cd *hivev1.ClusterDeployment, logger log.FieldLogger) error {
+	dnsZone, err := r.dnsZone(cd, logger)
+	if err != nil {
+		return errors.Wrap(err, "failed to get DNSZone")
+	}
+	objects := make([]hivev1.MetaRuntimeObject, 0, 2)
+	if dnsZone != nil {
+		objects = append(objects, dnsZone)
+	}
+	objects = append(objects, cd)
+
+	for _, obj := range objects {
+		if changed := controllerutils.ClearRelocateAnnotation(obj); !changed {
+			continue
+		}
+		if err := r.Update(context.Background(), obj); err != nil {
+			logger.WithError(err).Log(controllerutils.LogLevel(err), "failed to clear relocate annotation")
+			return errors.Wrap(err, "failed to clear relocate annotation")
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileClusterRelocate) finishRelocateCompletion(cd *hivev1.ClusterDeployment, relocateName string, logger log.FieldLogger) (reconcile.Result, error) {
+	if err := r.setRelocateAnnotation(cd, relocateName, hivev1.RelocateComplete, logger); err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "could not set relocate status to complete")
+	}
+
+	metricSuccessfulClusterRelocations.WithLabelValues(relocateName).Inc()
```

Comment:
are we risking multiple counting if errors are encountered below?

Reply:
Yes. This was a carry-over from when the controller would not reconcile the ClusterDeployment any more once the relocated annotation was set. That has been fixed now so that the controller will still attempt to delete the ClusterDeployment again if there was an error. In light of that, I have moved the metrics increment to after the delete.